active/0000-service-workers.md

```diff
@@ -0,0 +1,48 @@
+- Start Date: 2018-03-13
+- RFC PR: (leave this empty)
+
+# Add a Service Worker with minimal asset caching to a default Ember CLI application
+## Summary
+This RFC introduces a Service Worker to the default app blueprint. It will only cache the most essential assets, such as `index.html`, and the vendor and app specific JS and CSS files.
+
+## Motivation
+A Service Worker allows you to cache assets and other resources in a  way that loading a page becomes reliable and fast, independent of what the network conditions are.
+
+Not all Ember applications are destined to become a Progressive Web App, but that doesn't mean that applications shouldn't leverage a simple Service Worker script to somewhat improve the reliability of loading it.
+
+## Detailed design
+### Registration
+A Service Worker script named `sw.js` should be compiled and placed at the root level of the build result. The Service Worker script should be registered with a registration script that is inlined in `index.html`. The registered Service Worker should have the `rootURL` that is configured in `config/environment.js` as scope.
+
+### Caching
+The default Service Worker configuration should only cache the following files:
+
+* `index.html`
+* `assets/vendor.js`
+* `assets/<app-name>.js`
+* `assets/vendor.css`
+* `assets/<app-name>.css`
+
+The cached `index.html` file should be served by the Service Worker whenever a `fetch` event happens that has the request mode of `navigation` and the `accept` header of the request has `text/html` as value.
+
+### Service Worker invalidation
+By default each build should include a random sequence of text in the compiled output of `sw.js`, this way the Service Worker gets invalidated after every build.
+
+### Kill switch
+The build process should be able to take a flag that acts as a kill switch in case the Service Worker needs to be unregistered in the deployed environment as the result of a bad deploy.
```

Comment:
If we aim to make a killswitch part of the default setup, it's essential that we make its limitations and side-effects known.
- A potentially unintuitive consequence of SW unregister: push subscriptions will be invalidated. 
- There are also some situations where a simple unregister is not enough to dislodge a misbehaving worker (i.e., someone has passed a forever-pending promise to `ExtendableEvent#waitUntil`)
- What happens if worker v16 is installed in a user's browser, and they're not online when the killswitch is flipped. Are we forcing all apps down a path where `skipWaiting()` must be called? It's not desirable to deviate from the standard serviceworker lifecycle in some situations.

A robust kill switch is **not** trivial to build, and I have concerns about giving developers a false sense of safety by stating that this requirement is handled.

Reply:
^^ Probably the most important comment on the thread.